primary/tests/integration_tests_configuration_api.rs

```diff
@@ -142,3 +141,224 @@ fn connect_to_configuration_client(parameters: Parameters) -> ConfigurationClien
         .unwrap();
     ConfigurationClient::new(channel)
 }
+
+#[tokio::test]
+async fn epoch_change() {
+    let parameters = Parameters {
+        batch_size: 200, // Two transactions.
+        ..Parameters::default()
+    };
+
+    // The configuration of epoch 0.
+    let keys_0 = keys(None);
+    let committee_0 = pure_committee_from_keys(&keys_0);
+
+    // Spawn the committee of epoch 0.
+    let mut rx_channels = Vec::new();
+    let mut tx_channels = Vec::new();
+    for keypair in keys_0 {
+        let name = keypair.public().clone();
+        let signer = keypair;
+
+        let (tx_new_certificates, rx_new_certificates) = channel(CHANNEL_CAPACITY);
+        rx_channels.push(rx_new_certificates);
+        let (tx_feedback, rx_feedback) = channel(CHANNEL_CAPACITY);
+        tx_channels.push(tx_feedback);
+
+        let store = NodeStorage::reopen(temp_dir());
+
+        Primary::spawn(
+            name,
+            signer,
+            Arc::new(committee_0.clone()),
+            parameters.clone(),
+            store.header_store.clone(),
+            store.certificate_store.clone(),
+            store.payload_store.clone(),
+            /* tx_consensus */ tx_new_certificates,
+            /* rx_consensus */ rx_feedback,
+            /* dag */ None,
+            NetworkModel::Asynchronous,
+        );
+    }
+
+    // Run for a while in epoch 0.
+    for rx in rx_channels.iter_mut() {
+        loop {
+            let certificate = rx.recv().await.unwrap();
+            assert_eq!(certificate.epoch(), 0);
+            if certificate.round() == 10 {
+                break;
+            }
+        }
+    }
+
+    // Move to the next epochs.
+    for epoch in 1..=3 {
+        // Move to the next epoch.
+        let new_committee = committee_0.clone();
+        new_committee.epoch.swap(Arc::new(epoch));
+        for tx in &tx_channels {
+            tx.send(ConsensusPrimaryMessage::Committee(new_committee.clone()))
+                .await
+                .unwrap();
+        }
+
+        // Run for a while.
+        for rx in rx_channels.iter_mut() {
+            loop {
+                let certificate = rx.recv().await.unwrap();
+                if certificate.epoch() == epoch && certificate.round() == 10 {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+// #[tokio::test]
+// async fn partial_committee_change() {
```

Comment:
I surmise that the difficulty you've been having is in the repeated use of `committee_1` in `for tx in &epoch_0_tx_channels`... the committee indeed is an empty shell around two `ArcSwap`s, which you can refer to for cheap (according to `Arc` semantics), but you'll have to re-create the empty shell (the `Committee` itself) which is consumed by sending it in a message.

In the grand scheme of things, I would not try to address this directly. Rather, I would :
1. address the remark on `config/lib.rs` that rather than have an `ArcSwap<Epoch>` and an `ArcSwap<BTreeMap<...>>`, we'd rather have a single `ArcSwap` containing both,
2. notice we do not need those fancy containers in a `Message` type. Indeed: the `Arc...` are here to manage internal single-node concurrency of a piece of data that's shared across many tasks on the local node. Over the wire, we only need to share flat data, those fancy containers created at node A have no particular meaning for another node B (B != A). This points to having a `CommitteeMessage` type that's analogous to the old variant of the `Committee`, i.e. : 
```rust
CommitteeMessage {
  authorities: BTreeMap<PublicKey, Authority>,
  epoch: Epoch
}
```
and simple implementations for `From<CommitteeMessage> for Committee` and `From<&Commitee> for CommitteeMessage` that help with the translation.

Reply:
I see! thanks!